cisco_ise/actions/lib/base.py

```diff
@@ -0,0 +1,458 @@
+            Not required if the action handles its own parsing of params to get URI/Data
+        """
+
+        # Basic housekeeping for running the action.
+        # Creates self.action_kwargs and others
+        self._prepare_kwargs(**kwargs)
+
+        # Setup the connection details and parse Config Vs Action parameters.
+        # Creates self.connection
+        self._setup_connection(CONNECTION_FIELDS)
+
+        # Perpare the URL string to use with the request.
+        # Creates to self.request_uri
+        self._prepare_uri()
+
+        # Return True incase something cares to check
+        return True
+
+    def api_client(self,
+                   http_method=None,
+                   request_uri=None,
+                   api_user=None,
+                   api_pass=None,
+                   verify_ssl=True,
+                   payload=None,
+                   payload_format=None):
+
+        """ Parse data, prepare the request, and call the api_request()
+            This API Client holds your hand through request process.
+            If you just want to send things directly, use api_request()
+        """
+
+        # Check that the HTTP method we are attempting to use is in the
+        # list of valid methods (HTTP_METHODS)
+        if http_method not in HTTP_METHODS:
+            raise KeyError(
+                'http_method is a required parameter for "requests" and must be one of '
+                f'"{HTTP_METHODS}". Method Supplied: {http_method}')
+
+        if not request_uri:
+            self.logger.error(
+                'api_request(request_uri) is required, but none provided')
+            sys.exit(1)
+
+        # Check if formatting details were passed
+        # if not, check if self.payload_format exists and use that value
+        # If it doesn't exist, just set it to JSON
+        if not payload_format:
+            try:
+                self.payload_format
+            except NameError:
+                payload_format = 'json'
+            else:
+                payload_format = self.payload_format
+
+        # Validate payload_format
+        if payload_format not in ['json', 'xml']:
+            self.logger.error(
+                f'Valid types for payload_format are ["json", "xml"]. Received "{payload_format}"')
+            sys.exit(1)
+
+        headers = {
+            'Accept': 'application/json',
+            'Content-Type': f'application/{payload_format}'
+        }
+
+        auth = HTTPBasicAuth(api_user, api_pass)
+
+        # GET and DELETE don't need a payload sent
+        # Sometimes a GET will be a 'Get-All' which is effectively a search.
+        # In that case, results need to be handled variably
+        if http_method in ['GET', 'DELETE']:
+            result = []
+            last = False
+
+            # Run the api_request() on a loop until we have no more pages
+            # page['SearchResult']['nextPage']
+            while True:
```

Comment:
You could do `while not last:` here and then move the `return` block outside of the while loop.

Reply:
I agree, that would be more explicit and easier to understand.